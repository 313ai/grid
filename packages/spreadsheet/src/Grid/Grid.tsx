import React, { useCallback, useRef, useMemo, useEffect } from "react";
import Grid, {
  RendererProps,
  useSelection,
  GridRef,
  useEditable,
  useCopyPaste,
  useUndo,
  CellInterface,
  SelectionProps,
  SelectionArea,
  ScrollCoords,
} from "@rowsncolumns/grid";
import { debounce } from "@rowsncolumns/grid/dist/helpers";
import {
  ThemeProvider,
  ColorModeProvider,
  useTheme,
  useColorMode,
} from "@chakra-ui/core";
import {
  COLUMN_HEADER_WIDTH,
  DEFAULT_COLUMN_WIDTH,
  ROW_HEADER_HEIGHT,
  DEFAULT_ROW_HEIGHT,
  DARK_MODE_COLOR_LIGHT,
} from "./../constants";
import HeaderCell from "./../HeaderCell";
import Cell from "./../Cell";
import { GridWrapper, ThemeType } from "./../styled";
import { Cells } from "../Spreadsheet";

export interface SheetGridProps {
  theme?: ThemeType;
  minColumnWidth?: number;
  minRowHeight?: number;
  rowCount?: number;
  columnCount?: number;
  CellRenderer?: React.FC<RendererProps>;
  HeaderCellRenderer?: React.FC<RendererProps>;
  width?: number;
  height?: number;
  cells: Cells;
  onChange: (changes: Cells) => void;
  activeCell: CellInterface | null;
  selections: SelectionArea[];
  onSheetChange: (props: any) => void;
  selectedSheet: string;
  onScroll: (state: ScrollCoords) => void;
  scrollState: ScrollCoords;
}

export interface RowColSelection {
  rows: number[];
  cols: number[];
}

/**
 * Grid component
 * @param props
 */
const SheetGrid: React.FC<SheetGridProps> = (props) => {
  const {
    theme,
    minColumnWidth = DEFAULT_COLUMN_WIDTH,
    minRowHeight = DEFAULT_ROW_HEIGHT,
    rowCount = 1000,
    columnCount = 1000,
    width,
    height,
    CellRenderer = Cell,
    HeaderCellRenderer = HeaderCell,
    cells,
    onChange,
    onSheetChange,
    activeCell: initialActiveCell,
    selections: initialSelections = [],
    selectedSheet,
    onScroll,
    scrollState,
  } = props;
  const gridRef = useRef<GridRef | null>(null);
  const { colorMode } = useColorMode();
  const isLightMode = colorMode === "light";
  const onSheetChangeRef = useRef(debounce(onSheetChange, 300));

  /**
   * Get cell value or text
   */
  const getValue = useCallback(
    ({ rowIndex, columnIndex }: CellInterface, obj = false) => {
      return rowIndex in cells
        ? obj
          ? cells[rowIndex][columnIndex]
          : cells[rowIndex][columnIndex]?.text
        : undefined;
    },
    [cells]
  );

  /**
   * Selection
   */
  const {
    selections,
    activeCell,
    setActiveCell,
    setSelections,
    ...selectionProps
  } = useSelection({
    initialActiveCell,
    initialSelections,
    gridRef,
    rowCount,
    columnCount,
    onFill: (activeCell, fillSelection, allSelections) => {
      if (!fillSelection) return;
      /* Check if user is trying to extend a selection */
      const { bounds } = fillSelection;
      const changes: Cells = {};
      const previousChanges = {};
      const value = getValue(activeCell, true);
      for (let i = bounds.top; i <= bounds.bottom; i++) {
        for (let j = bounds.left; j <= bounds.right; j++) {
          if (!(i in changes)) changes[i] = {};
          changes[i][j] = value;
        }
      }
      onChange?.(changes);
    },
  });

  /**
   * If grid changes, lets restore the state
   */
  useEffect(() => {
    if (scrollState) gridRef.current?.scrollTo(scrollState);
    setActiveCell(initialActiveCell);
    setSelections(initialSelections);
  }, [selectedSheet]);

  /**
   * Save it back to sheet
   */
  useEffect(() => {
    onSheetChangeRef.current?.({ activeCell, selections });
  }, [activeCell, selections]);

  const handleSubmit = useCallback(
    (
      value: string,
      cell: CellInterface,
      nextActiveCell?: CellInterface | null
    ) => {
      const { rowIndex, columnIndex } = cell;
      const changes = {
        [rowIndex]: {
          [columnIndex]: {
            text: value,
          },
        },
      };

      onChange?.(changes);

      /* Focus on next active cell */
      if (nextActiveCell) setActiveCell(nextActiveCell);
    },
    [cells]
  );

  /**
   * Editable
   */
  const { editorComponent, isEditInProgress, ...editableProps } = useEditable({
    gridRef,
    selections,
    activeCell,
    onSubmit: handleSubmit,
    getValue,
  });

  /**
   * Copy paste
   */
  useCopyPaste({
    gridRef,
    getValue,
    selections,
  });
  /**
   * Undo/redo
   */
  const { undo, redo, add, canUndo, canRedo, ...undoProps } = useUndo();

  /* Width calculator */
  const columnWidth = useCallback(
    (columnIndex: number) => {
      if (columnIndex === 0) return COLUMN_HEADER_WIDTH;
      return minColumnWidth;
    },
    [minColumnWidth]
  );
  const rowHeight = useCallback(
    (rowIndex: number) => {
      if (rowIndex === 0) return ROW_HEADER_HEIGHT;
      return minRowHeight;
    },
    [minColumnWidth]
  );
  const contextWrapper = useCallback(
    (children) => {
      return (
        <ThemeProvider theme={theme}>
          <ColorModeProvider>{children}</ColorModeProvider>
        </ThemeProvider>
      );
    },
    [theme]
  );
  const selectedRowsAndCols: RowColSelection = useMemo(
    () =>
      selections.reduce(
        (acc, { bounds }) => {
          for (let i = bounds.left; i <= bounds.right; i++) {
            acc.cols.push(i);
          }
          for (let i = bounds.top; i <= bounds.bottom; i++) {
            acc.rows.push(i);
          }
          return acc;
        },
        { rows: [], cols: [] } as RowColSelection
      ),
    [selections]
  );

  const itemRenderer = useCallback(
    (props: RendererProps) => {
      const { rowIndex, columnIndex } = props;
      const isRowHeader = rowIndex === 0;
      const isColumnHeader = columnIndex === 0;
      const isHeaderActive =
        isRowHeader || isColumnHeader
          ? isRowHeader
            ? activeCell?.columnIndex === columnIndex ||
              selectedRowsAndCols.cols.includes(columnIndex)
            : activeCell?.rowIndex === rowIndex ||
              selectedRowsAndCols.rows.includes(rowIndex)
          : false;
      const cell = { rowIndex, columnIndex };
      if (rowIndex === 0 || columnIndex === 0)
        return <HeaderCellRenderer {...props} isActive={isHeaderActive} />;
      return <CellRenderer {...props} {...getValue(cell, true)} />;
    },
    [CellRenderer, HeaderCellRenderer, cells, selectedRowsAndCols, activeCell]
  );
  const fillhandleBorderColor = isLightMode ? "white" : DARK_MODE_COLOR_LIGHT;

  return (
    <GridWrapper>
      <Grid
        ref={gridRef}
        rowCount={rowCount}
        columnCount={columnCount}
        columnWidth={columnWidth}
        rowHeight={rowHeight}
        frozenColumns={1}
        frozenRows={1}
        width={width}
        height={height}
        itemRenderer={itemRenderer}
        wrapper={contextWrapper}
        selections={selections}
        activeCell={activeCell}
        fillhandleBorderColor={fillhandleBorderColor}
        showFillHandle={!isEditInProgress}
        {...selectionProps}
        {...editableProps}
        onScroll={(scrollState: ScrollCoords) => {
          editableProps.onScroll(scrollState);
          onScroll(scrollState);
        }}
        onKeyDown={(e: React.KeyboardEvent<HTMLDivElement>) => {
          selectionProps.onKeyDown(e);
          editableProps.onKeyDown(e);
          undoProps.onKeyDown(e);
        }}
      />
      {editorComponent}
    </GridWrapper>
  );
};

export default SheetGrid;
