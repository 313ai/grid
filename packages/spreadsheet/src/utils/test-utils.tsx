import React from "react";
import { render } from "@testing-library/react";
import { Stage, Layer } from "react-konva";
import { ThemeProvider, theme } from "@chakra-ui/core";
import Konva from "konva";

const KonvaWrapper = ({ children }) => {
  return (
    <Stage>
      <ThemeWrapper>
        <Layer>{children}</Layer>
      </ThemeWrapper>
    </Stage>
  );
};

const konvaRenderer = (ui) => render(ui, { wrapper: KonvaWrapper });
const domRenderer = (ui) => render(ui, { wrapper: ThemeWrapper });

export * from "@testing-library/react";
export { konvaRenderer, domRenderer };

export const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
