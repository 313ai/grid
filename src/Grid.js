"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const react_konva_1 = require("react-konva");
const defaultProps = {
    width: 800,
    height: 800,
    rowCount: 200,
    columnCount: 200,
    rowHeight: () => 20,
    columnWidth: () => 100,
    scrollbarSize: 20,
};
const getRowStartIndex = (rowCount, rowHeight, scrollTop) => {
    let i = 0;
    let startIndex = 0;
    while (i < rowCount) {
        const itemHeight = rowHeight(i);
        if (i * itemHeight + itemHeight >= scrollTop) {
            startIndex = i;
            break;
        }
        i++;
    }
    return startIndex;
};
const getRowStopIndex = (startIndex, rowCount, rowHeight, scrollTop, containerHeight) => {
    let i = startIndex;
    let stopIndex = rowCount;
    while (i < rowCount) {
        const itemHeight = rowHeight(i);
        const itemScrollTop = i * itemHeight - scrollTop;
        if (itemScrollTop > containerHeight) {
            stopIndex = i;
            break;
        }
        i++;
    }
    return stopIndex;
};
const getColumnStartIndex = (columnCount, columnWidth, scrollLeft) => {
    let i = 0;
    let startIndex = 0;
    while (i < columnCount) {
        const itemWidth = columnWidth(i);
        if (i * itemWidth + itemWidth >= scrollLeft) {
            startIndex = i;
            break;
        }
        i++;
    }
    return startIndex;
};
const getColumnStopIndex = (startIndex, columnCount, columnWidth, scrollLeft, containerWidth) => {
    let i = startIndex;
    let stopIndex = columnCount;
    while (i < columnCount) {
        const itemWidth = columnWidth(i);
        const itemScrollLeft = i * itemWidth - scrollLeft;
        if (itemScrollLeft > containerWidth) {
            stopIndex = i;
            break;
        }
        i++;
    }
    return stopIndex;
};
const getBoundedCells = (area) => {
    const { top, bottom, left, right } = area;
    const cells = new Set();
    for (let i = top; i <= bottom; i++) {
        for (let j = left; j <= right; j++) {
            cells.add(JSON.stringify([i, j]));
        }
    }
    return cells;
};
const itemKey = ({ rowIndex, columnIndex }) => `${rowIndex}:${columnIndex}`;
/**
 * Grid component
 * @param props
 */
const Grid = (props) => {
    const { width: containerWidth, height: containerHeight, rowHeight, columnWidth, rowCount, columnCount, scrollbarSize, children, } = props;
    const verticalScrollRef = react_1.useRef(null);
    const wheelingRef = react_1.useRef(null);
    const horizontalScrollRef = react_1.useRef(null);
    const [scrollTop, setScrollTop] = react_1.useState(0);
    const [scrollLeft, setScrollLeft] = react_1.useState(0);
    const handleScroll = react_1.useCallback((e) => {
        setScrollTop(e.target.scrollTop);
    }, []);
    const handleScrollLeft = react_1.useCallback((e) => {
        setScrollLeft(e.target.scrollLeft);
    }, []);
    const scrollHeight = rowCount * rowHeight();
    const scrollWidth = columnCount * columnWidth();
    const [selectedArea, setSelectedArea] = react_1.useState({
        top: 0,
        bottom: 0,
        left: 0,
        right: 0,
    });
    const boundedCells = react_1.useMemo(() => getBoundedCells(selectedArea), [
        selectedArea,
    ]);
    const handleWheel = react_1.useCallback((event) => {
        if (wheelingRef.current)
            return;
        const { deltaX, deltaY, deltaMode } = event.nativeEvent;
        let dx = deltaX;
        let dy = deltaY;
        if (deltaMode === 1) {
            dy = dy * 17;
        }
        if (!horizontalScrollRef.current || !verticalScrollRef.current)
            return;
        const x = horizontalScrollRef.current?.scrollLeft;
        const y = verticalScrollRef.current?.scrollTop;
        wheelingRef.current = window.requestAnimationFrame(() => {
            wheelingRef.current = null;
            if (horizontalScrollRef.current)
                horizontalScrollRef.current.scrollLeft = x + dx;
            if (verticalScrollRef.current)
                verticalScrollRef.current.scrollTop = y + dy;
        });
    }, []);
    const rowStartIndex = getRowStartIndex(rowCount, rowHeight, scrollTop);
    const rowStopIndex = getRowStopIndex(rowStartIndex, rowCount, rowHeight, scrollTop, containerHeight);
    const columnStartIndex = getColumnStartIndex(columnCount, columnWidth, scrollLeft);
    const columnStopIndex = getColumnStopIndex(columnStartIndex, columnCount, columnWidth, scrollLeft, containerWidth);
    const items = [];
    if (columnCount > 0 && rowCount) {
        for (let rowIndex = rowStartIndex; rowIndex <= rowStopIndex; rowIndex++) {
            for (let columnIndex = columnStartIndex; columnIndex <= columnStopIndex; columnIndex++) {
                const width = columnWidth(columnIndex);
                const x = columnIndex * width;
                const height = rowHeight(rowIndex);
                const y = rowIndex * height;
                items.push(react_1.createElement(children, {
                    x,
                    y,
                    width,
                    height,
                    rowIndex,
                    columnIndex,
                    key: itemKey({ rowIndex, columnIndex }),
                }));
            }
        }
    }
    return (react_1.default.createElement("div", { style: { position: "relative", width: containerWidth + 20 } },
        react_1.default.createElement("div", { style: {
                height: containerHeight,
                overflow: "scroll",
                position: "absolute",
                right: 0,
                top: 0,
                width: scrollbarSize,
                background: "#666",
            }, onScroll: handleScroll, ref: verticalScrollRef },
            react_1.default.createElement("div", { style: {
                    position: "absolute",
                    height: scrollHeight,
                    width: 1,
                } })),
        react_1.default.createElement("div", { style: {
                overflow: "scroll",
                position: "absolute",
                bottom: -scrollbarSize,
                left: 0,
                width: containerWidth,
                height: scrollbarSize,
                background: "#666",
            }, onScroll: handleScrollLeft, ref: horizontalScrollRef },
            react_1.default.createElement("div", { style: {
                    position: "absolute",
                    width: scrollWidth,
                    height: 1,
                } })),
        react_1.default.createElement("div", { onWheel: handleWheel, tabIndex: -1 },
            react_1.default.createElement(react_konva_1.Stage, { width: containerWidth, height: containerHeight },
                react_1.default.createElement(react_konva_1.Layer, null,
                    react_1.default.createElement(react_konva_1.Group, { offsetY: scrollTop, offsetX: scrollLeft }, items))))));
};
Grid.defaultProps = defaultProps;
exports.default = Grid;
