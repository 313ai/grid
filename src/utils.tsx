import React from "react";
import { Line, Rect } from "react-konva";
import { ShapeConfig } from "konva/types/Shape";

interface BoxProps extends ShapeConfig {
  onCrosshairMouseDown?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseMove?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseUp?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Create a box with custom top/right/bottom/left colors and widths
 * @param param0
 */
export const createBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
}: BoxProps) => {
  const commonProps = {
    perfectDrawEnabled: false,
    shadowForStrokeEnabled: false,
    hitStrokeWidth: 0,
    listening: false,
    lineCap: "square",
  };
  const lines = [
    <Line
      points={[x, y, x + width, y]}
      stroke={strokeTopColor}
      strokeWidth={strokeTopWidth}
      key="top"
      {...commonProps}
    />,
    <Line
      points={[x + width, y, x + width, y + height]}
      stroke={strokeRightColor}
      strokeWidth={strokeRightWidth}
      key="right"
      {...commonProps}
    />,
    <Line
      points={[x + width, y + height, x, y + height]}
      stroke={strokeBottomColor}
      strokeWidth={strokeBottomWidth}
      key="bottom"
      {...commonProps}
    />,
    <Line
      points={[x, y + height, x, y]}
      stroke={strokeLeftColor}
      strokeWidth={strokeLeftWidth}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <Rect
          fill={fill}
          x={x}
          y={y}
          width={width}
          height={height}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};

export const FillHandle: React.FC<BoxProps> = ({
  x = 0,
  y = 0,
  stroke,
  onMouseDown,
}) => {
  if (x === 0 || y === 0) return null;
  const crossHairSize = 8;
  const crossHairBorderSize = 1;
  return (
    <div
      style={{
        position: "absolute",
        left: x - crossHairSize / 2,
        top: y - crossHairSize / 2,
        width: crossHairSize,
        height: crossHairSize,
        border: `${crossHairBorderSize}px white solid`,
        background: stroke,
        cursor: "crosshair",
      }}
      onMouseDown={onMouseDown}
    />
  );
};

export const createHTMLBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
  strokeStyle = "solid",
}: BoxProps) => {
  const commonProps = {};
  width = width - Math.floor(strokeWidth);
  y = y - Math.ceil(strokeWidth / 2);
  const lines = [
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: width,
        height: strokeTopWidth,
        borderWidth: 0,
        borderColor: strokeTopColor,
        borderTopWidth: strokeTopWidth,
        borderStyle: strokeStyle,
      }}
      key="top"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x + width,
        top: y,
        width: strokeRightWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeRightColor,
        borderRightWidth: strokeRightWidth,
        borderStyle: strokeStyle,
      }}
      key="right"
      {...commonProps}
    />,
    ,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y + height,
        width: width + strokeTopWidth,
        height: strokeBottomWidth,
        borderWidth: 0,
        borderColor: strokeBottomColor,
        borderBottomWidth: strokeBottomWidth,
        borderStyle: strokeStyle,
      }}
      key="bottom"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: strokeLeftWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeLeftColor,
        borderLeftWidth: strokeLeftWidth,
        borderStyle: strokeStyle,
      }}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <div
          style={{
            position: "absolute",
            top: y,
            left: x,
            height,
            width,
            backgroundColor: fill,
            userSelect: "none",
            pointerEvents: "none",
          }}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};
