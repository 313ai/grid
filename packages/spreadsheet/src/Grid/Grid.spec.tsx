import React from "react";
import { domRenderer, cleanup } from "./../utils/test-utils";
import Grid from "./Grid";
import { theme } from '@chakra-ui/core'

describe("Grid", () => {
  afterEach(cleanup);
  it("renders a spreadsheet grid", () => {
    const renderGrid = () =>
      domRenderer(
        <Grid
          cells={{
            1: {
              1: {
                text: 'hello'
              }
            }
          }}
          activeCell={null}
          selections={[]}
          selectedSheet={null}    
          theme={theme}
        />
      );
    expect(renderGrid).not.toThrow();
  });

  it("matches snapshot", () => {
    const { asFragment } = domRenderer(
      <Grid
        cells={{
          1: {
            1: {
              text: 'hello'
            }
          }
        }}
        activeCell={null}
        selections={[]}
        selectedSheet={null}
        theme={theme}      
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
})