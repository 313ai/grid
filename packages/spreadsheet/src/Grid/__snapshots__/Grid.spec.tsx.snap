// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-oqtn58"
  >
    <div
      class="rowsncolumns-grid"
      style="position: relative; width: 800px; user-select: none;"
    >
      <div
        class="rowsncolumns-grid-container"
        tabindex="0"
      >
        <div>
          <div
            class="konvajs-content"
            role="presentation"
            style="position: relative; user-select: none; width: 800px; height: 600px;"
          >
            <canvas
              height="600"
              style="padding: 0px; margin: 0px; border: 0px; background: transparent; position: absolute; top: 0px; left: 0px; width: 800px; height: 600px; display: block;"
              width="800"
            />
          </div>
        </div>
      </div>
      <div
        style="pointer-events: none;"
      >
        <div
          style="position: absolute; left: 46px; top: 24px; right: 0px; bottom: 0px; overflow: hidden;"
        >
          <div
            style="transform: translate(-46px, -24px);"
          />
        </div>
        <div
          style="position: absolute; width: 54px; top: 24px; left: 0px; bottom: 0px; overflow: hidden;"
        >
          <div
            style="transform: translate(0, -24px);"
          />
        </div>
        <div
          style="position: absolute; height: 32px; left: 46px; right: 0px; top: 0px; overflow: hidden;"
        >
          <div
            style="transform: translate(-46px, 0);"
          />
        </div>
        <div
          style="position: absolute; height: 32px; width: 54px; left: 0px; top: 0px; overflow: hidden; pointer-events: none;"
        />
      </div>
      <div
        class="rowsncolumns-grid-scrollbar rowsncolumns-grid-scrollbar-y"
        style="height: 600px; overflow: scroll; position: absolute; right: 0px; top: 0px; width: 13px; will-change: transform;"
        tabindex="-1"
      >
        <div
          style="position: absolute; height: 49212px; width: 1px;"
        />
      </div>
      <div
        class="rowsncolumns-grid-scrollbar rowsncolumns-grid-scrollbar-x"
        style="overflow: scroll; position: absolute; bottom: 0px; left: 0px; width: 800px; height: 13px; will-change: transform;"
        tabindex="-1"
      >
        <div
          style="position: absolute; width: 50446px; height: 1px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
